Subject: [PATCH] return to OF via trap, not exit
From: Olaf Hering <olh@suse.de>
X-Git-Commit: add60ef303809087999412209d24d400a1c96670
Patch-mainline: 2.6.17-rc1

Do not call prom exit prom_panic. It clears the screen and the exit
message is lost.

On some (or all?) pmacs it causes another crash when OF tries to print
the date and time in its banner.

Set of_platform earlier to catch more prom_panic() calls.

Signed-off-by: Olaf Hering <olh@suse.de>
Acked-by: Segher Boessenkool <segher@kernel.crashing.org>
Signed-off-by: Paul Mackerras <paulus@samba.org>

 arch/powerpc/kernel/prom_init.c |   23 ++++++++++++++---------
 1 file changed, 14 insertions(+), 9 deletions(-)

Index: linux-2.6.16/arch/powerpc/kernel/prom_init.c
===================================================================
--- linux-2.6.16.orig/arch/powerpc/kernel/prom_init.c
+++ linux-2.6.16/arch/powerpc/kernel/prom_init.c
@@ -397,6 +397,11 @@ static void __init __attribute__((noretu
 	reason = PTRRELOC(reason);
 #endif
 	prom_print(reason);
+	/* Do not call exit because it clears the screen on pmac
+	 * it also causes some sort of double-fault on early pmacs */
+	if (RELOC(of_platform) == PLATFORM_POWERMAC)
+		asm("trap\n");
+
 	/* ToDo: should put up an SRC here on p/iSeries */
 	call_prom("exit", 0, 0);
 
@@ -2061,15 +2066,6 @@ unsigned long __init prom_init(unsigned 
 	 */
 	prom_init_stdout();
 
-	/* Bail if this is a kdump kernel. */
-	if (PHYSICAL_START > 0)
-		prom_panic("Error: You can't boot a kdump kernel from OF!\n");
-
-	/*
-	 * Check for an initrd
-	 */
-	prom_check_initrd(r3, r4);
-
 	/*
 	 * Get default machine type. At this point, we do not differentiate
 	 * between pSeries SMP and pSeries LPAR
@@ -2079,6 +2075,15 @@ unsigned long __init prom_init(unsigned 
 	prom_setprop(_prom->chosen, "/chosen", "linux,platform",
 		     &getprop_rval, sizeof(getprop_rval));
 
+	/* Bail if this is a kdump kernel. */
+	if (PHYSICAL_START > 0)
+		prom_panic("Error: You can't boot a kdump kernel from OF!\n");
+
+	/*
+	 * Check for an initrd
+	 */
+	prom_check_initrd(r3, r4);
+
 #ifdef CONFIG_PPC_PSERIES
 	/*
 	 * On pSeries, inform the firmware about our capabilities
